compiler/rustc_hir_typeck/src/method/probe.rs

```diff
@@ -1484,20 +1486,29 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
             Option<ObligationCause<'tcx>>,
         )>,
     ) -> ProbeResult {
-        debug!("consider_probe: self_ty={:?} probe={:?}", self_ty, probe);
-
         self.probe(|_| {
             // First check that the self type can be related.
-            let sub_obligations = match self.at(&ObligationCause::dummy(), self.param_env).sup(
-                DefineOpaqueTypes::No,
-                probe.xform_self_ty,
-                self_ty,
-            ) {
-                Ok(InferOk { obligations, value: () }) => obligations,
-                Err(err) => {
-                    debug!("--> cannot relate self-types {:?}", err);
+            let sub_obligations = match self_ty.kind() {
+                // HACK: opaque types will match anything for which their bounds hold.
+                // Thus we need to prevent them from trying to match the `&_` autoref
+                // candidates that get created for `&self` trait methods.
```

Comment:
Not just `&self`, right? Also `&mut self`?

Also, I kind of expect that if the self-type is an opaque that can be defined we would just give an error if it hasn't been constrained yet. 

Do I have an inaccurate mental model?

Reply:
Even unconstrained, users are relying on this already, and since we can support it, it seems odd to reject it.

We do support `&self` and all other receivers (mutable, pin, ...), as long as the hidden type is constrained to `Self` and not a reference to it. There's tests for that.